File: LeetCode/LeetCode/OtherActivities/AlgorithmStudyPlan/Algorithm2/Problem74.cs
using System.Linq;

namespace LeetCode.OtherActivities.AlgorithmStudyPlan.Algorithm2
{
    public class Problem74
    {
        public bool SearchMatrix(int[][] matrix, int target)
        {
            if (matrix == null || matrix.Length == 0)
                return false;
            var i = matrix.Length;
            var j = matrix[0].Length;

            var left = 0;
            var right = i * j - 1;
            while (true)
            {
                var mid = (right + left) / 2;
                var midI = mid / j;
                var midJ = mid % j;
                if (matrix[midI][midJ] == target)
                    return true;

                if (left > right)
                    return false;

                if (matrix[midI][midJ] > target)
                {
                    right = mid - 1;
                    
                }
                else
                {
                    left = mid + 1;
                }
            }
        }
    }
}

File: leetcode_rust/src/problems/problem_74.rs
pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
    let height = matrix.len();
    let width = matrix[0].len();
    let mut left = 0usize;
    let mut right = height * width - 1;

    while right - left > 1 {
        let mid = (right + left) / 2;
        let mid_el = matrix[mid / width][mid % width];
        if mid_el == target {
            return true;
        }
        if mid_el > target { right = mid; } else { left = mid; }
    }

    if matrix[left / width][left % width] == target || matrix[right / width][right % width] == target {
        true
    } else {
        false
    }
}
